### PR TITLE
Add Autopilot feature from Consul

### PR DESCRIPTION
**Pull request by kyhavlov**
This adds the OSS components for autopilot as well as the config fields/docs for the enterprise autopilot features. The autopilot logic is in the agent/consul/autopilot package over in the Consul repo.

### CODE REVIEW

File: nomad/util.go [hunk 1]
```diff
@@ -80,6 +83,12 @@ func isNomadServer(m serf.Member) (bool, *serverParts) {
 		}
 	}
 
+	// If the server is missing the rpc_addr tag, default to the serf advertise addr
+	rpc_ip := net.ParseIP(m.Tags["rpc_addr"])
```
Comment (schmichael) [inline nomad/util.go]: rpcIP would be a more idiomatic name

File: website/source/guides/cluster/autopilot.html.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+```
+{
+    "cleanup_dead_servers": true,
+    "last_contact_threshold": "200ms",
+    "max_trailing_logs": 250,
+    "server_stabilization_time": "10s",
+    "redundancy_zone_tag": "az",
+    "disable_upgrade_migration": false,
+    "upgrade_version_tag": ""
+}
```
Comment (schmichael) [inline website/source/guides/cluster/autopilot.html.md]: HCLify?

File: website/source/guides/cluster/autopilot.html.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+servers as soon as a replacement Nomad server comes online. When servers are removed
+by the cleanup process they will enter the "left" state.
+
+This option can be disabled by running `nomad operator autopilot set-config`
+with the `-cleanup-dead-servers=false` option.
+
+## Server Health Checking
+
+An internal health check runs on the leader to track the stability of servers.
+</br>A server is considered healthy if all of the following conditions are true:
```
Comment (schmichael) [inline website/source/guides/cluster/autopilot.html.md]: `<br>` not `</br>` but is it necessary?

File: nomad/server.go [hunk 4]
```diff
@@ -271,6 +279,9 @@ func NewServer(config *Config, consulCatalog consul.CatalogAPI, logger *log.Logg
 	// Create the periodic dispatcher for launching periodic jobs.
 	s.periodicDispatcher = NewPeriodicDispatch(s.logger, s)
 
+	// Initialize the stats fetcher that autopilot will use.
+	s.statsFetcher = NewStatsFetcher(logger, s.connPool, s.config.Region)
+
 	// Setup Vault
 	if err := s.setupVaultClient(); err != nil {
```
Comment (preetapan) [inline nomad/server.go]: The `Leave` method here also calls RemovePeer and currently does not handle raft protocol 3, I left a TODO in there in the master branch.

Comment (preetapan) [reply nomad/server.go]: Another TODO https://github.com/hashicorp/nomad/blob/master/api/operator.go#L72

File: command/agent/agent.go [hunk 5]
```diff
@@ -160,6 +160,32 @@ func convertServerConfig(agentConfig *Config, logOutput io.Writer) (*nomad.Confi
 	if agentConfig.Sentinel != nil {
 		conf.SentinelConfig = agentConfig.Sentinel
 	}
+	if agentConfig.Server.NonVotingServer {
+		conf.NonVoter = true
+	}
+	if agentConfig.Autopilot != nil {
+		if agentConfig.Autopilot.CleanupDeadServers != nil {
+			conf.AutopilotConfig.CleanupDeadServers = *agentConfig.Autopilot.CleanupDeadServers
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply command/agent/agent.go]: I didn't want to override the defaults in the server config with different defaults from the user config

File: nomad/server.go (hunk 4 above)
Comment (kyhavlov) [reply nomad/server.go]: The last of these (raft ID related) should be taken care of now

**Comment by kyhavlov** [comment]
@schmichael those references to datacenter instead of region were just mistakes, I think.

This should be ready for another look now.

File: command/operator_autopilot.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	return cli.RunResultHelp
+}
+
+func (c *OperatorAutopilotCommand) Synopsis() string {
+	return "Provides tools for modifying Autopilot configuration"
+}
+
+func (c *OperatorAutopilotCommand) Help() string {
+	helpText := `
+Usage: consul operator autopilot <subcommand> [options]
```
Comment (schmichael) [inline command/operator_autopilot.go]: consul -> nomad

File: command/operator_autopilot_get.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	return 0
+}
+
+func (c *OperatorAutopilotGetCommand) Synopsis() string {
+	return "Display the current Autopilot configuration"
+}
+
+func (c *OperatorAutopilotGetCommand) Help() string {
+	helpText := `
+Usage: consul operator autopilot get-config [options]
```
Comment (schmichael) [inline command/operator_autopilot_get.go]: consul -> nomad

File: command/operator_autopilot_set.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	return 1
+}
+
+func (c *OperatorAutopilotSetCommand) Synopsis() string {
+	return "Modify the current Autopilot configuration"
+}
+
+func (c *OperatorAutopilotSetCommand) Help() string {
+	helpText := `
+Usage: consul operator autopilot set-config [options]
```
Comment (schmichael) [inline command/operator_autopilot_set.go]: consul -> nomad

File: command/operator_autopilot_set.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+  Modifies the current Autopilot configuration.
+
+General Options:
+
+  ` + generalOptionsUsage() + `
+
+Set Config Options:
+
+  -cleanup-dead-servers=[true|false]
+     Controls whether Consul will automatically remove dead servers when
```
Comment (schmichael) [inline command/operator_autopilot_set.go]: consul -> nomad

File: command/operator_autopilot_set.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+     new ones are successfully added. Must be one of [true|false].
+
+  -disable-upgrade-migration=[true|false]
+     (Enterprise-only) Controls whether Consul will avoid promoting
```
Comment (schmichael) [inline command/operator_autopilot_set.go]: consul -> nomad

File: command/operator_autopilot_set.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+  -server-stabilization-time=<10s>
+     Controls the minimum amount of time a server must be stable in
+     the 'healthy' state before being added to the cluster. Only takes
+     effect if all servers are running Raft protocol version 3 or
+     higher. Must be a duration value such as "10s".
+
+  -upgrade-version-tag=<value>
+     (Enterprise-only) The node_meta tag to use for version info when
+     performing upgrade migrations. If left blank, the Consul version
```
Comment (schmichael) [inline command/operator_autopilot_set.go]: consul -> nomad

File: nomad/structs/config/autopilot.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+	// servers into zones for redundancy. If left blank, this feature will be disabled.
+	RedundancyZoneTag string `mapstructure:"redundancy_zone_tag"`
+
+	// (Enterprise-only) DisableUpgradeMigration will disable Autopilot's upgrade migration
+	// strategy of waiting until enough newer-versioned servers have been added to the
+	// cluster before promoting them to voters.
+	DisableUpgradeMigration *bool `mapstructure:"disable_upgrade_migration"`
+
+	// (Enterprise-only) UpgradeVersionTag is the node tag to use for version info when
+	// performing upgrade migrations. If left blank, the Consul version will be used.
```
Comment (schmichael) [inline nomad/structs/config/autopilot.go]: consul -> nomad

File: website/source/guides/cluster/autopilot.html.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+## Redundancy Zones
+
+Prior to Autopilot, it was difficult to deploy servers in a way that took advantage of
+isolated failure domains such as AWS Availability Zones; users would be forced to either
+have an overly-large quorum (2-3 nodes per AZ) or give up redundancy within an AZ by
+deploying just one server in each.
+
+If the `RedundancyZoneTag` setting is set, Nomad will use its value to look for a
+zone in each server's specified [`-meta`](/docs/agent/configuration/client.html#meta)
```
Comment (dadgar) [inline website/source/guides/cluster/autopilot.html.md]: This is only defined on clients that aren't participating in Raft/Serf.

File: api/operator_autopilot.go [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+	if err != nil {
+		return err
+	}
+	*d = ReadableDuration(dur)
+	return nil
+}
+
+// AutopilotGetConfiguration is used to query the current Autopilot configuration.
+func (op *Operator) AutopilotGetConfiguration(q *QueryOptions) (*AutopilotConfiguration, error) {
+	r, err := op.c.newRequest("GET", "/v1/operator/autopilot/configuration")
```
Comment (dadgar) [inline api/operator_autopilot.go]: 152-166 can be replaced by:

```
var resp AutopilotConfiguration
qm, err := n.client.query("/v1/operator/autopilot/configuration", &resp, q)
if err != nil {
	return nil, nil, err
}
return resp, qm, nil
```

All of the APIs should use this format. There is a query as well as write helper for PUTs

File: api/operator_autopilot.go [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+	dur, err := time.ParseDuration(str[1 : len(str)-1])
+	if err != nil {
+		return err
+	}
+	*d = ReadableDuration(dur)
+	return nil
+}
+
+// AutopilotGetConfiguration is used to query the current Autopilot configuration.
+func (op *Operator) AutopilotGetConfiguration(q *QueryOptions) (*AutopilotConfiguration, error) {
```
Comment (dadgar) [inline api/operator_autopilot.go]: Should also return the query meta result: `(*AutopilotConfiguration, *QueryMeta, error)` and in the cases of writes, `WriteMeta`

File: command/agent/operator_endpoint.go [hunk 16]
```diff
@@ -49,21 +55,222 @@ func (s *HTTPServer) OperatorRaftPeer(resp http.ResponseWriter, req *http.Reques
...
-	if _, ok := params["address"]; ok {
-		args.Address = raft.ServerAddress(params.Get("address"))
-	} else {
+	_, hasID := params["id"]
+	_, hasAddress := params["address"]
+
+	if !hasID && !hasAddress {
+		resp.WriteHeader(http.StatusBadRequest)
+		fmt.Fprint(resp, "Must specify either ?id with the server's ID or ?address with IP:port of peer to remove")
+		return nil, nil
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: Change to be `return nil, CodedError(http.StatusBadRequest, error_string)` here and else where were this is used

File: command/agent/operator_endpoint.go [hunk 17]
```diff
@@ -49,21 +55,222 @@ func (s *HTTPServer) OperatorRaftPeer(resp http.ResponseWriter, req *http.Reques
...
+		}
+
+		// Only use the out value if this was a CAS
+		if !args.CAS {
+			return true, nil
+		}
+		return reply, nil
+
+	default:
+		resp.WriteHeader(http.StatusMethodNotAllowed)
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: return nil, CodedError(404, ErrInvalidMethod)

File: command/agent/operator_endpoint.go [hunk 18]
```diff
@@ -49,21 +55,222 @@ func (s *HTTPServer) OperatorRaftPeer(resp http.ResponseWriter, req *http.Reques
...
+			UpgradeVersionTag:       reply.UpgradeVersionTag,
+			CreateIndex:             reply.CreateIndex,
+			ModifyIndex:             reply.ModifyIndex,
+		}
+
+		return out, nil
+
+	case "PUT":
+		var args structs.AutopilotSetConfigRequest
+		s.parseRegion(req, &args.Region)
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: parseWriteRequest

File: nomad/operator_endpoint.go [hunk 19]
```diff
@@ -124,3 +125,161 @@ REMOVE:
...
+// interface.
+func (op *Operator) RaftRemovePeerByID(args *structs.RaftPeerByIDRequest, reply *struct{}) error {
+	if done, err := op.srv.forward("Operator.RaftRemovePeerByID", args, args, reply); done {
+		return err
+	}
+
+	// Check management permissions
+	if aclObj, err := op.srv.ResolveToken(args.AuthToken); err != nil {
+		return err
+	} else if aclObj != nil && !aclObj.IsManagement() {
```
Comment (dadgar) [inline nomad/operator_endpoint.go]: Why not OperatorWrite?

File: nomad/server.go [hunk 20]
```diff
@@ -777,6 +801,8 @@ func (s *Server) setupRPC(tlsWrap tlsutil.RegionWrapper) error {
 	}
 	s.rpcListener = list
 
+	s.logger.Printf("[INFO] nomad: RPC listening on %q", s.rpcListener.Addr().String())
```
Comment (dadgar) [inline nomad/server.go]: Is this useful? I would say lets remove

File: website/source/api/operator.html.md [hunk 21]
```diff
@@ -123,3 +125,232 @@ $ curl \
...
+The table below shows this endpoint's support for
+[blocking queries](/api/index.html#blocking-queries),
+[consistency modes](/api/index.html#consistency-modes), and
+[required ACLs](/api/index.html#acls).
+
+| Blocking Queries | Consistency Modes | ACL Required    |
+| ---------------- | ----------------- | --------------- |
+| `NO`             | `none`            | `operator:read` |
+
+### Parameters
```
Comment (dadgar) [inline website/source/api/operator.html.md]: I think we remove both of these parameters (dc should be region) but they are both covered generically on this page: https://www.nomadproject.io/api/#consistency-modes

File: website/source/api/operator.html.md [hunk 22]
```diff
@@ -123,3 +125,232 @@ $ curl \
...
+[consistency modes](/api/index.html#consistency-modes), and
+[required ACLs](/api/index.html#acls).
+
+| Blocking Queries | Consistency Modes | ACL Required     |
+| ---------------- | ----------------- | ---------------- |
+| `NO`             | `none`            | `opreator:write` |
+
+### Parameters
+
+- `dc` `(string: "")` - Specifies the datacenter to query. This will default to
```
Comment (dadgar) [inline website/source/api/operator.html.md]: Remove all of these across endpoints

File: website/source/docs/agent/configuration/autopilot.html.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+  The "autopilot" stanza configures the Nomad agent to configure Autopilot behavior.
+---
+
+# `autopilot` Stanza
+
+<table class="table table-bordered table-striped">
+  <tr>
+    <th width="120">Placement</th>
+    <td>
+      <code>**acl**</code>
```
Comment (dadgar) [inline website/source/docs/agent/configuration/autopilot.html.md]: acl -> autopilot

File: nomad/operator_endpoint.go (hunk 19 above)
Comment (kyhavlov) [reply nomad/operator_endpoint.go]: That seems fine to me; the rest of the raft methods in here take a management token though

**Pull request merged**